sphinx/util/typing.py

```diff
@@ -147,7 +149,63 @@ def is_system_TypeVar(typ: Any) -> bool:
     return modname == 'typing' and isinstance(typ, TypeVar)
 
 
-def restify(cls: type | None, mode: str = 'fully-qualified-except-typing') -> str:
+class RenderMode(enum.Flag):
+    """Additional flags for rendering annotations or reST content."""
+
+    # principal modes (mutually exclusive with each other)
+
+    smart = enum.auto()
+    """Show the annotation name."""
+
+    fully_qualified = enum.auto()
+    """Show the module name and qualified name of the annotation.
+
+    This mode is mutually exclusive with :attr:`smart` and :attr:`fully_qualified_except_typing`.
+    """  # NoQA: E501
+
+    fully_qualified_except_typing = enum.auto()
+    """Same as :attr:`fully_qualified` but do not show module name for ``typing`` members.
+
+    This mode is mutually exclusive with :attr:`smart` and :attr:`fully_qualified`.
+    """
+
+    # secondary modes (can be used as flags on the primary modes)
+
+    short_literal = enum.auto()
+    """Use PEP 604 style to render literals."""
+
+
+_MUTUALLY_EXCLUSIVE_MODES = list(itertools.combinations([
+    RenderMode.smart,
+    RenderMode.fully_qualified,
+    RenderMode.fully_qualified_except_typing,
+], 2))
+
+
+def _normalize_mode(mode: str | RenderMode) -> RenderMode:  # for backwards compatibility
+    if isinstance(mode, RenderMode):
+        for a, b in _MUTUALLY_EXCLUSIVE_MODES:
+            if mode & a and mode & b:
```

Comment:
Optional suggestion:

```suggestion
            if flag_a in mode and flag_b in mode:
```

Reply:
For flags, I will advocate for `&` and `|`.